extensions/amp-selector/0.2/selector.js

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as CSS from './selector.css';
+import * as Preact from '../../../src/preact';
+import {omit} from '../../../src/utils/object';
+import {useContext, useState} from '../../../src/preact';
+
+const SelectorContext = Preact.createContext({});
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function Selector(props) {
+  const {
+    'as': Comp = 'div',
+    'children': children,
+    'disabled': disabled,
+    'value': value,
+    'multiple': multiple,
+  } = props;
+  const [selectedState, setSelectedState] = useState(value ? value : []);
+  // TBD: controlled values require override of properties.
+  const selected = /** @type {!Array} */ (value ? value : selectedState);
+  const selectOption = (option) => {
+    if (!option) {
+      return;
+    }
+    const {'onChange': onChange} = props;
+    let newValue = null;
+    if (multiple) {
+      newValue = selected.includes(option)
+        ? selected.filter((v) => v != option)
+        : selected.concat(option);
+    } else if (!selected.includes(option)) {
+      newValue = [option];
+    }
+    if (newValue) {
+      setSelectedState(newValue);
```

Comment:
This appears to be a reasonable change in the Preact component but not the AMP component because the `value` prop is always set to the initial state, which for many of our examples is `[]`.

Reply:
I think we should go with Preact on this level. And see what's the best mapping on AMP's level. But I don't see the full context here yet. Is there some particular situation in AMP that doesn't work with this?

Reply:
Am I understanding correctly that you want to update the `if (newValue)` condition to the `if (value == null)` condition? If this is the case then for AMP mode initializes the `value` with `[]`, not `null`, meaning it never goes down this `setSelectedState` path. 

I also tried setting the AMP mode `value` to `null` when it resolves to `[]`, but this only work until the first item is selected, at which point `value` becomes non-empty and is unable to go down this code path again. 

Perhaps we need a different prop to convey the AMP state other than `value`?

Reply:
Btw, it's ok to defer this item. Just add a tracking item to https://github.com/ampproject/amphtml/issues/28282. 

I think it's `if (value != null)`, right? In React these things are controlled usually with two props: `value` and `defaultValue`. And there're two modes:

- `value != null` means that the caller controls all of the state. We can duplicate this step internally, but that could have lots of problems for us - just a bit fragile and always a place to make a bug in. This is a controlled component. I think the safest way for us to go is to not use state at all.
- `value == null` - our component controls the state. This is an uncontrolled component.

I think in the training, we even replaced `setStateValue` callback with a no-op or indirected function to avoid setting it accidentally.